### PR TITLE
jupyter-web-app: Integrate with PodDefaults

### LINKED ISSUE #2992: Jupyter spawner ui - let user select admission hooks to apply to the notebook
We are increasingly using webhooks to inject volumes and environment variables into Jupyter pods.

For example

  * On GCP we inject GCP credentials into Jupyter pods
  * We'd like to use to inject GIT secrets (see #2889 )

To support this we'd like the jupyter spawner UI to let users choose which admission hooks to apply to their notebooks.

Admission hooks are controlled based on labels. So step 1 would be to let users specify which labels to apply to their notebooks.

The next step would allow the web ui to introspect the cluster to see which admission hooks are installed in the cluster. Ideally each admission hook would have a human readable label describing what it does (e.g. as annotation). This would be something like

   * Attach S3 credentials for buckets ...

#2844 we'd like to generalize the admission web hook so users could easily define admission hooks for different things. Possibly by creating a CR similar to PodPresets.

Then the spawner UI could show a set of drop down boxes with these descriptions and a user could just check it. Under the hood the spawner UI could get the labels used by the corresponding admission hook and attach them. This will require some UI work as well as work on generalizing the web admission hook #2844.

/cc @zabbasi @vkoukis @kimwnasptd 

@zabbasi might pick up some of this work. I think a design doc would be a good place to start.

### LINKED ISSUE #3536: jupyter-web-app: Error 200 on load
In the new UI if no namespace is selected from the centraldashboard before opening the Jupyter UI, then the value `""` is passed as namespace. Because of this the GET will hit an invalid endpoint resulting in an error 200 popup.

The webapp should ignore the empty value from the centraldashboard.

### PR DESCRIPTION
**Pull request by kimwnasptd**
closes #2992 
closes #3536 

The user can select which PodDefaults will be applied in the Notebooks Pod. In the UI's `select` element the user sees the `desc` field of the `PodDefaults`. 

Under the hood the UI will attach the labels for the selected PodDefaults to the new Pods, which will trigger the `PodDefault` controller.

Also, the `config.yaml` has a [configurations](https://github.com/kubeflow/manifests/blob/master/jupyter/jupyter-web-app/base/config-map.yaml#L134) section from which an admin can set a default set of labels - PodDefaults that will be selected from the UI on load. The admin could also make the `readOnly` attr in the config `true` to block the users from editing this section.

@jlewi @zabbasi In the UI I've named the section as `Configurations`. Should we instead maintain the 1:1 naming with the k8s objects and rename it to `Pod Defaults`?
configs

/cc zabbasi
/cc prodonjs
/cc jlewi

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/3537)

### CODE REVIEW

**Comment by jlewi** [comment]
/assign @lluunn 
/assign @avdaredevil 

@avdaredevil Could you review the JS/Web changes please?

@lluunn Could you review functionality/admission hooks integration please?

File: components/jupyter-web-app/backend/kubeflow_jupyter/common/api.py [hunk 1]
```diff
@@ -101,6 +101,17 @@ def get_notebooks(ns):
     )
 
 
+def get_poddefaults(ns):
+    return wrap_resp(
+        'poddefaults',
```
Comment (lluunn) [inline components/jupyter-web-app/backend/kubeflow_jupyter/common/api.py]: just curious: the pattern here seems to be using single quote for first param, and double quote for third and later ones. What's the reason?

File: components/jupyter-web-app/backend/kubeflow_jupyter/common/utils.py [hunk 2]
```diff
@@ -334,6 +334,23 @@ def set_notebook_memory(notebook, body, defaults):
...
+        labels = defaults["configurations"]["value"]
+        logger.info("Using default Configurations: {}".format(labels))
+    elif body.get("configurations", None) is not None:
+        labels = body["configurations"]
+        logger.info("Using form's Configurations: {}".format(labels))
+    else:
+        labels = defaults["configurations"]["value"]
+        logger.info("Using default Configurations: {}".format(labels))
+
+    for l in labels:
```
Comment (lluunn) [inline components/jupyter-web-app/backend/kubeflow_jupyter/common/utils.py]: do we need to check the labels is iterable?

**Comment by lluunn** [comment]
Just minor questions. Backend LGTM

 I think we call it "configuration" might be confusing, both in FE and BE.

**Comment by jlewi** [comment]
@kimwnasptd Our priority is unblocking the release. So if we have something that is good enough (sounds like we do) lets go with what we have and we can consider future improvements later.

If its an easy fix then someone just make the call about what to go with and lets do it.

**Comment by jlewi** [comment]
@kimwnasptd Do you have an ETA for when we can get this PR merged and the package in kubeflow/manifests updated?

**Comment by kimwnasptd** [comment]
@lluunn @avdaredevil updated the code to align with your comments

@jlewi I think we should go with `configurations` in this PR to unblock the release.

There are still some open issues regarging the webapp and multi user implementation like #2980 and #2981. I'd like to put priority on them first. Once we are done with them I can come up and rename to `Pod Defaults`.

If we won't have enough time before the release to make this change, then we could keep `Configurations` and I'll update the docs to make it clear that this section refers to `Pod Defaults`. Then changing it to `Pod Defaults` will be one of my first PRs for 0.7.

Also, I had already updated the `config.yaml` in the manifests so after we merge this PR we would only need to update the image.

WDYT?

**Comment by jlewi** [comment]
Sounds good. We can create minor releases 0.6.x too if needed. 

Lets try to do the following ASAP

1. Get this PR merged
1. Get the image on master updated
1. Deploy on master and see what's working.

Specifically with regards to #2980. Does spawning notebooks work on non GCP?

e.g. do things work on minikf?

**Comment by kimwnasptd** [comment]
Not in the `kubeflow` Namespace. The Notebooks that get created from the webapp will use the `default-editor` SA. If the creation of a Namespace happens from the Profile Controller then this SA will be created and the Notebooks Pod will be deployed.

But this SA is not present in the `kubeflow` Namespace. As you [said](https://github.com/kubeflow/kubeflow/issues/3502#issuecomment-503828467) we could create a profile for the `kubeflow` Namespace or we could create the `default-editor` SA from the webapp's manifests for `kubeflow`. Either way, lets open an issue to discuss/keep track of it there.

**Comment by kimwnasptd** [comment]
It seems that the `kubeflow-presubmit-tf-notebooks-release` test is failing. IIRC this one builds the images for the Notebooks.

Checked the commits for `components/tensorflow-notebook-image` and found this [commit](https://github.com/kubeflow/kubeflow/pull/3587/files) from two days ago that updated the Dockerfile. 

Also, from the [logs](http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-presubmit-tf-notebook-release-3537-416def6-7024-25bb?tab=workflow&nodeId=kubeflow-presubmit-tf-notebook-release-3537-416def6-7024-25bb-851366601&sidePanel=logs%3Akubeflow-presubmit-tf-notebook-release-3537-416def6-7024-25bb-851366601%3Amain) I can see that it fails when it tries to 
```bash
/bin/bash -c pip install --upgrade pip==19.0.1 && pip --no-cache-dir install ${TF_PACKAGE}.... ${PIPELINE_SDK_PACKAGE} ....
```

@IronPan could the new version of the Pipeline SDK be the problem or it might something else?

**Comment by jlewi** [comment]
@kimwnasptd Can you try rebasing on the latest master?
I don't think this PR should be triggering the notebook presubmit if it didn't modify that code. 
Its probably doing that because you haven't picked up the latest changes from master so it things this PR is modifying the code in the notebook images.

**Comment by jlewi** [comment]
/lgtm
/approve
/hold

I'm rubber stamping the PR. If the PR is good to go and the tests pass just remove the hold.

**Pull request merged**